scripts/get-all-images.py

```diff
@@ -0,0 +1,223 @@
+    except subprocess.CalledProcessError as exc:
+        logging.error("Script '%s' for charm '%s' failed: %s",
+                      GET_IMAGES_SH, app["charm"], exc.stderr)
+        raise exc
+    finally:
+        # cleanup
+        shutil.rmtree(repo_dir)
+
+    images = process.stdout.split("\n")
+
+    return images
+
+
+def get_dependency_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a dependency charm by:
+    1. Cloning the repo of the charm
+    2. Looking at its metadata.yaml for "upstream-source" keys
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def cleanup_images(images: list[str]) -> list[str]:
+    """
+    Given a list of OCI registry images ensure
+    1. there are no duplicates
+    2. there are no images with empty name
+    3. the list is sorted
+
+    Args:
+        images: List of images to be processed
+
+    Returns:
+        A new list without duplicate values, empty values and is sorted
+    """
+    images_set = set(images)
+    if '' in images_set:
+        images_set.remove('')
+
+    unique_images = list(images_set)
+    unique_images.sort()
+
+    return unique_images
+
+
+def get_bundle_images(bundle_path: str) -> list[str]:
+    """Return a list of images used by a bundle"""
+    bundle_dict = {}
+    with open(bundle_path, 'r') as file:
+        bundle_dict = yaml.safe_load(file)
+
+    images = []
+
+    for app_name, app in bundle_dict["applications"].items():
+        logging.info(f"Handling app {app_name}")
+
+        # exclude repos we know don't have images
+        if app_name in EXCLUDE_CHARMS:
+            logging.info("Ignoring charm %s", app["charm"])
+            continue
+
+        # Follow default image-gather process for dependency apps
+        if is_dependency_app(app):
+            logging.info("Dependency app '%s' with charm '%s'", app_name,
+                         app["charm"])
+            images.extend(get_dependency_app_images(app))
+            continue
+
+        # Ensure analytics app has necessary github repo/branch metadata
+        if not bundle_app_contains_gh_metadata(app):
+            raise KeyError(
+                "Application '%s' doesn't include expected gh metadata keys" %
+                app_name
+            )
+
+        images.extend(get_analytics_app_images(app))
+
+    return cleanup_images(images)
+
+
+if __name__ == "__main__":
```

Comment:
Just a nit, usually the more pythonic way is to have a `main()` method and call it, instead of adding the logic directly here.

```
if __name__ == "__main__":
  main()
```

Reply:
done!
61ddca8